### PR TITLE
add go.mod and go.sum support

### PR DESCRIPTION
**Pull request by calebdoxsey**
For: https://github.com/Microsoft/vscode-go/issues/1886

I was 90% done with this before I realized there was already a PR out there from a few months ago. If this isn't useful, feel free to close.

### CODE REVIEW

**Comment by ramya-rao-a** [comment]
Thanks for the PR @calebdoxsey. Looks like this is your first contribution to this project, Thanks & Welcome!

> I realized there was already a PR out there from a few months ago

Which PR are you referring to here?

**Comment by calebdoxsey** [comment]
Here's the PR in the atom repo: https://github.com/atom/language-go/pull/156

**Comment by ramya-rao-a** [comment]
@brainsnail Do you know if there is an ETA for your PR (https://github.com/atom/language-go/pull/156)?

If not, then would you be interested in reviewing this PR and give us your feedback?

**Comment by brainsnail** [comment]
@ramya-rao-a As far as I know there hasn't been an update on that PR, but I'm fine with taking a look at this PR instead 😄

I'll take a look at it this soon @calebdoxsey.

File: syntaxes/go.sum.tmGrammar.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+							"match": ".*",
+							"name": "invalid.illegal.unknown-hash.go.sum"
+						}
+					]
+				}
+			}
+		},
+		"semver": {
+			"comment": "Semver version strings (v1.2.3)",
+			"match": "v(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)(?:-[\\da-z-]+(?:\\.[\\da-z-]+)*)?(?:\\+[\\da-z-]+(?:\\.[\\da-z-]+)*)?",
```
Comment (OneOfOne) [inline syntaxes/go.sum.tmGrammar.json]: the "[official](https://github.com/golang/go/blob/df557fe3ea94ff8888abd6a6ab4cc5d5351d77a6/src/cmd/go/internal/modfetch/pseudo.go#L89)" regexp is: `^v[0-9]+\.(0\.0-|\d+\.\d+-([^+]*\.)?0\.)\d{14}-[A-Za-z0-9]+(\+incompatible)?$`

Comment (calebdoxsey) [reply syntaxes/go.sum.tmGrammar.json]: I think that's for the special pseudo version: https://tip.golang.org/cmd/go/#hdr-Pseudo_versions

Comment (brainsnail) [reply syntaxes/go.sum.tmGrammar.json]: I think I agree about using that regex instead - I wrote one by hand on the Atom repo, but we may want to just use the same regex as the one the language uses.

**Comment by brainsnail** [comment]
I tried it out with a few different repos that had modules enabled and did some `go mod init`-ing on a few others to see how it handled the highlighting. As far as that goes - I think it works great. It highlighted the proper keywords on the `go.mod` files, gave appropriate highlighting on versions and kept comment formatting. The `go.sum` file highlights the most important piece - the version.

Super glad to see it moving forward 👍

File: package.json [hunk 2]
```diff
@@ -72,6 +72,36 @@
...
+          "Go Module File"
+        ]
+      },
+      {
+        "id": "go.sum",
+        "extensions": [
+          "go.sum"
+        ],
+        "aliases": [
+          "Go Sum File"
```
Comment (brainsnail) [inline package.json]: Can we call this `Go Checksum File`? It's a minor nitpick.

File: syntaxes/go.sum.tmGrammar.json (hunk 1 above)
Comment (calebdoxsey) [reply syntaxes/go.sum.tmGrammar.json]: The issue is it only matches pseudo versions: 

```v0.0.0-20161113214103-89cd22812c4f```

It won't match 

```v1.2.3```

At least that's how I read the regex. Is there another one that matches both?

File: package.json (hunk 2 above)
Comment (calebdoxsey) [reply package.json]: Yep. Good idea.

File: syntaxes/go.sum.tmGrammar.json (hunk 1 above)
Comment (brainsnail) [reply syntaxes/go.sum.tmGrammar.json]: I think you are right - the code they use to validate the semver is [here](https://github.com/golang/go/blob/aab0b704d88b7dad10c767ad941619d59887203c/src/cmd/go/internal/semver/semver.go#L150). 

In testing out your branch locally I didn't experience any problems.

**Comment by ramya-rao-a** [comment]
Thanks a lot @brainsnail and @calebdoxsey!

**Pull request merged**